mkosi

```diff
@@ -2204,42 +2249,125 @@ def install_arch(args: CommandLineArguments, root: str, do_run_build_script: boo
                     )
                 )
 
-    def stop_gpg_agent() -> None:
-        # Kill the gpg-agent used by pacman and pacman-key
-        run(['gpgconf', '--homedir', os.path.join(root, 'etc/pacman.d/gnupg'), '--kill', 'all'])
+    if not do_run_build_script and args.bootable:
+        hooks_dir = os.path.join(root, "etc/pacman.d/hooks")
+        scripts_dir = os.path.join(root, "etc/pacman.d/scripts")
+
+        os.makedirs(hooks_dir, 0o755, exist_ok=True)
+        os.makedirs(scripts_dir, 0o755, exist_ok=True)
+
+        # Disable depmod pacman hook as depmod is handled by kernel-install as well.
+        os.symlink("/dev/null", os.path.join(hooks_dir, "60-depmod.hook"))
+
+        kernel_add_hook = os.path.join(hooks_dir, "kernel-add.hook")
+        with open(kernel_add_hook, 'w') as f:
+            f.write(
+                dedent(
+                    """\
+                    [Trigger]
+                    Operation = Install
+                    Operation = Upgrade
+                    Type = Path
+                    Target = usr/lib/modules/*/vmlinuz
+
+                    [Trigger]
+                    Operation = Install
+                    Operation = Upgrade
+                    Operation = Remove
+                    Type = Package
+                    Target = systemd
+                    Target = dracut
+                    Target = intel-ucode
+                    Target = amd-ucode
+
+                    [Action]
+                    Description = Adding kernel and initramfs images to /boot...
+                    When = PostTransaction
+                    Exec = /etc/pacman.d/scripts/kernel-add
+                    NeedsTargets
+                    """
+                )
+            )
+
+        kernel_add_script = os.path.join(scripts_dir, "kernel-add")
+        with open(kernel_add_script, 'w') as f:
+            f.write(
+                dedent(
+                    """\
+                    #!/bin/bash -e
+
+                    ALL=0
+
+                    read -r -d '' -a TARGETS || true
+                    for i in "${!TARGETS[@]}"; do
+                        [[ ${TARGETS[$i]} != */vmlinuz ]] && ALL=1
+                        TARGETS[$i]="/${TARGETS[$i]}"
+                    done
+
+                    [[ "$ALL" ]] && TARGETS=(/usr/lib/modules/*/vmlinuz)
+
+                    for f in "${TARGETS[@]}"; do
+                        kernel-install add $(basename $(dirname $f)) $f
+                    done
```

Comment:
Thinking about it, this works, but, as bash is, it's a bit noisy. On my personal, I have something similar to this, which is a bit more verbose, but hopefully has less line noise
```bash
#!/bin/bash -e                                                                                                   
                                                                                                                 
declare kernel_version                                                                                           
                                                                                                                 
while read -r line                                                                                               
do                                                                                                               
    if [[ "$line" =~ usr/lib/modules/([^/]+)/vmlinuz ]]                                                          
    then                                                                                                         
        kernel_version="${BASH_REMATCH[1]}"                                                                      
        break                                                                                                    
    fi                                                                                                           
done                                                                                                             
                                                                                                                 
                                                                                                                 
if [[ -n "$kernel_version" ]]                                                                                    
then                                                                                                             
    kernel-install add "$kernel_version" "/usr/lib/${kernel_version}/vmlinuz"                                    
else                                                                                                             
    for f in /usr/lib/modules/*/vmlinuz                                                                          
    do                                                                                                           
        kernel-install add $(basename $(dirname $f)) $f                                                          
    done                                                                                                         
fi       
```

Reply:
Does this work when 2 kernels out of 5 (for example) get upgraded?

Reply:
No, this works under the assumption that the update will be either a kernel package or one of the others. This can be changed to
```bash
#!/bin/bash -e

declare -a kernel_version

# Check lines we are passed from the pacman hook
while read -r line
do
    if [[ "$line" =~ usr/lib/modules/([^/]+)/vmlinuz ]]
    then
        kernel_version+=("${BASH_REMATCH[1]}")
    fi
done

# If kernel_version is empty, the hook was triggered by an update, that was not
# a kernel package
if [[ "${#kernel_version[@]}" -eq 0 ]]
then
    for f in /usr/lib/modules/*/vmlinuz
    do
        kernel_version+=( $(basename $(dirname $f)) )
    done
fi
                                                                                                                                                                                                                                   
# (re)build the kernel images
for kv in "${kernel_version[@]}"
do
    kernel-install add "$kv" "/usr/lib/modules/${kv}/vmlinuz"
done
```
which I think is still more readable, but also more verbose (although it can be compressed, I like the idiosyncratic linebreaks in the syntax that are usually omitted via a semicolon)

Reply:
All of these scripts also have in common, that they would break, if somebody ever got the idea of having a space in `/usr/lib/modules/*/vmlinuz`.

Reply:
I'm still not sure this does exactly what we want. With this version, if dracut and a kernel are updated at the same time, only a single kernel will be rebuilt no?

Looking again at this I do see that my ALL boolean doesn't work. I pushed a new version with that addressed.

Reply:
Ah, true. Missed that edge case.
```bash
#!/bin/bash -e

declare -a kernel_version

# Check lines we are passed from the pacman hook
while read -r line
do
    if [[ "$line" =~ usr/lib/modules/([^/]+)/vmlinuz ]]
    then
        kernel_version+=( "${BASH_REMATCH[1]}" )
    else
        # If a non-matching line is passed, just rebuild all kernels
        kernel_version=()
        for f in /usr/lib/modules/*/vmlinuz
        do
            kernel_version+=( "$(basename "$(dirname "$f")")" )
        done
        break
    fi
done

# (re)build the kernel images
for kv in "${kernel_version[@]}"
do
    kernel-install add "$kv" "/usr/lib/modules/${kv}/vmlinuz"
done
```
It's unfortunate, that there is nothing cleaner than die basename dirname combo. 

While I like mine better (less sigil noise), as this point it's in the eye of the beholder. :)

Reply:
I'll honor all the effort you put in by picking your solution. I do agree that it's easier to read without all the sigil noise.

Reply:
No worries, a couple fewer edge cases for my work machine. :) (Not that it matters a lot, since it always ever has a single kernel.)